source/Tool/Plugins.cpp

```diff
@@ -0,0 +1,63 @@
+//
+// Copyright (c) 2023 Klemens Morgenstern (klemens.morgenstern@gmx.net)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+
+#include "Plugins.hpp"
+
+#include <mrdox/Config.hpp>
+#include <mrdox/Generators.hpp>
+#include <mrdox/Plugin.hpp>
+#include <mrdox/Support/Path.hpp>
+#include <mrdox/Support/Report.hpp>
+#include <mrdox/Version.hpp>
+
+#include "Support/Path.hpp"
+#include "Support/Debug.hpp"
+
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/DynamicLibrary.h>
+
+namespace clang {
+namespace mrdox {
+
+Error
+loadPlugins(
+        std::string_view addonsDir
+        )
+{
+
+    std::string err_;
+
+    SmallPathString pluginPath{addonsDir};
+    pluginPath.append("/plugins");
+
+    std::error_code ec;
+    for (auto itr = llvm::sys::fs::directory_iterator(pluginPath, ec);
+         itr != llvm::sys::fs::directory_iterator() && !ec;
+         itr.increment(ec))
+    {
+        if (!itr->path().ends_with(".dll")
+         && !itr->path().ends_with(".so"))
+            continue;
+
+        auto lib = llvm::sys::DynamicLibrary::getPermanentLibrary(itr->path().c_str(), &err_);
+        if (!lib.isValid())
+            return Error("couldn't load {} because {}", itr->path(), err_);
```

Comment:
Well no this message is not correct because the caller will end up printing this:
```
Could not initialize plugins because couldn't load {} because {}"
```

The text in the `Error` is the "because" so you might write this instead:

```
return Error( "getPermanentLibrary(\"{}\") returned \"{}\"" );
```